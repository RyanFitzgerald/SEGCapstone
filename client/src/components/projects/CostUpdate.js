import React from 'react';
import { Redirect } from 'react-router-dom';
import * as api from '../../api';

class CostUpdate extends React.Component {
  constructor() {
    super();

    // Bind functions
    this.handleSubmit = this.handleSubmit.bind(this);
    this.addUpdate = this.addUpdate.bind(this);
    this.getCents = this.getCents.bind(this);

    // Set state
    this.state = {
      redirect: null,
      formError: false
    };
  }

  componentDidMount() {
    // Update tab
    this.props.setActiveSubtab(0);

    // Set title
    document.title = 'Add Price Update | Renovaction';
  }

  handleSubmit(e) {
    // Stop form submission
    e.preventDefault();

    // Get form data
    const update = {
      amount: this.getCents(this.amount.value),
      reason: this.reason.value,
      type: this.updateType.value,
      project: this.props.location.match.params.id,
      addedBy: JSON.parse(localStorage.getItem('user'))._id,
      access_token: JSON.parse(localStorage.getItem('user')).access_token
    };

    // Call api
    this.addUpdate(update);
  }

  addUpdate(update) {
    api.addUpdate(update).then(resp => {
      if (resp.status === 500) {
        this.setState({
          formError: 'There was an error when submitting the form, please try again.'
        })
        return;
      }

      let contractCost = this.props.location.location.query.contractCost;

      if (update.type === 'Addition') {
        contractCost = contractCost + update.amount;
      } else {
        contractCost = contractCost - update.amount;
      }

      const project = {
        contractCost,
        access_token: JSON.parse(localStorage.getItem('user')).access_token
      }

      api.updateProject(project, this.props.location.match.params.id).then(resp => {
        if (resp.status === 500) {
          this.setState({
            formError: 'There was an error when submitting the form, please try again.'
          })
          return;
        }
  
        // Update parent state
        this.props.getProjects({search: false});
  
        // Redirect
        this.setState({
          redirect: {
            location: `/projects/${resp._id}`,
            message: 'Successfully updated project!',
            type: 'success'
          }
        });
      });     
    });
  }

  getCents(dollars) {
    return Math.round((Math.round(dollars * 100) / 100)*100);
  }
  
  render() {
    const id = this.props.location.match.params.id;
    let fileNumber = false;
    let salesPrice = false;

    if (this.props.location.location && this.props.location.location.query) {
      fileNumber = this.props.location.location.query.fileNumber;
      salesPrice = this.props.location.location.query.salesPrice;
    }

    if (this.state.redirect) {
      this.props.addNotification(this.state.redirect.message, this.state.redirect.type);
      return (
        <Redirect to={this.state.redirect.location} />
      );
    }
    
    if (id && fileNumber && salesPrice) {
      return (
        <div className="content">
          <div className="row">
            <div className="column">
              <h2 className="card-title">Add Price Update For <b>{fileNumber}</b></h2>
              <div className="card">
              {this.props.renderError(this.state.formError)}
                <form onSubmit={this.handleSubmit}>
                  <div className="row">
                    <div className="md-4 column form-section__title no-left">
                      <h3>Cost Update</h3>
                      <p>
                        Cost updates change the contract cost displayed for a project and a history of them is maintained on the project's page
                      </p>
                    </div>
                    <div className="md-8 column no-right">
                      <label className="form-label" htmlFor="name">Amount <span className="form-required">*</span></label>
                      <input name="amount" ref={input => this.amount = input} className="form-text form-text--full" type="number" step="0.01" required/>
                      <label className="form-label" htmlFor="reason">Reason <span className="form-required">*</span></label>
                      <input name="reason" ref={input => this.reason = input} className="form-text form-text--full" type="text" required/>
                      <label className="form-label" htmlFor="type">Type <span className="form-required">*</span></label>
                      <span className="form-select">
                        <select name="type" ref={input => this.updateType = input} required>
                          <option value="Addition">Addition</option>
                          <option value="Subtraction">Subtraction</option>
                        </select>
                      </span>
                    </div>
                  </div>
                  <div className="text-center">
                    <input type="submit" className="btn btn--primary btn--large" value="Add Update"/>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      );
    }
    
    return (
      <Redirect to="/projects"/>
    );
  }
}

export default CostUpdate;