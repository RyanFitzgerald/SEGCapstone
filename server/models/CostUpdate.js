/*
  Cost Update Model
  - This file contains the model for adding cost updates to a project
*/

const mongoose = require('mongoose');
mongoose.Promise = global.Promise;

const costUpdateSchema = new mongoose.Schema({
  created: {
    type: Date,
    default: Date.now
  },
  amount: {
    type: Number,
    required: 'An amount must be provided!'
  },
  reason: {
    type: String,
    required: 'A reason must be provided!',
    trim: true
  },
  type: {
    type: String,
    required: 'A type must be provided!'
  },
  project: {
    type: mongoose.Schema.ObjectId,
    ref: 'Project',
    required: 'A project must be provided!'
  },
  /*user: {
    type: mongoose.Schema.ObjectId,
    ref: 'User',
    required: 'A user must be provided!'
  }*/
});

module.exports = mongoose.model('CostUpdate', costUpdateSchema);