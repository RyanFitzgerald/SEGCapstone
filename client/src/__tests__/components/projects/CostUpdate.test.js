import React from 'react';
import CostUpdate from '../../../Components/projects/CostUpdate';

describe('<CostUpdate />', () => {
  let location;
  
  beforeAll(() => {
    // Create location prop with needed data
    location = {
      match: {
        params: {
          id: 1
        }
      }
    };
  });

  it('should render correctly', () => {   
    // Create mock functions / spies
    const setActiveSubtabMock = jest.fn();
    const renderErrorMock = jest.fn();
    
    // Shallow render it
    const wrapper = shallow(
      <CostUpdate setActiveSubtab={setActiveSubtabMock} renderError={renderErrorMock} location={location}/>
    );

    // Compare to old snapshot
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    // Check set active subtab is called and with correct value
    expect(setActiveSubtabMock).toHaveBeenCalledTimes(1);
    expect(setActiveSubtabMock.mock.calls[0][0]).toBe(0);
  });
});
