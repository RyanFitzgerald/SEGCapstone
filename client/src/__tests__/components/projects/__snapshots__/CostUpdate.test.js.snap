// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CostUpdate /> should render correctly 1`] = `
<Redirect
  push={false}
  to="/projects"
/>
`;
